src/pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -489,7 +497,7 @@ class _DataLoaderSource:
     instance: Optional[Union[TRAIN_DATALOADERS, EVAL_DATALOADERS, "pl.LightningModule", "pl.LightningDataModule"]]
     name: str
 
-    def dataloader(self) -> Union[TRAIN_DATALOADERS, EVAL_DATALOADERS]:
+    def dataloader(self) -> Union[TypeAlias["instance"], TRAIN_DATALOADERS, EVAL_DATALOADERS]:
```

Comment:
what is `TypeAlias["instance"]` supposed to be?

Reply:
without TypeAlias I get the following error:

```sh
src/pytorch_lightning/trainer/connectors/data_connector.py:500: error: Variable "pytorch_lightning.trainer.connectors.data_connector._DataLoaderSource.instance" is not valid as a type  [valid-type]
src/pytorch_lightning/trainer/connectors/data_connector.py:500: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

Reply:
The issue here seems that both of the types (`self.instance` as well as `self.dataloader()`) are valid, but mypy has a bit of a trouble here, since it thinks that it could happen, that `self.instance` is `pl.LightningModule` or `pl.LightningDataModule` and we return that value. Which is not true, as evidenced by the method, but mypy just doesn't understand it. In this case, I don't think there's an easy way to fix this and I'd go with `# type: ignore`

Any other opinions, @awaelchli @carmocca?

Reply:
I created a constant `INSTANCE = Optional[Union[TRAIN_DATALOADERS, EVAL_DATALOADERS, "pl.LightningModule", "pl.LightningDataModule"]]` and updated the return to `Union[INSTANCE, TRAIN_DATALOADERS, EVAL_DATALOADERS]`; this works locally, updated in https://github.com/Lightning-AI/lightning/pull/13806/commits/1aebec4fe5f7cb291784e2164367c60a578b8f79

Reply:
to recreate the original error, I removed the return type in https://github.com/Lightning-AI/lightning/pull/13806/commits/b0cb0409a05469990f4805b953ed937f06707bd3

Reply:
reverted in https://github.com/Lightning-AI/lightning/pull/13806/commits/dfca1fe56087253e8fc3fb2bab52b7bdcce72377